### PR TITLE
refactor: migrate to TypeScript

### PR DESCRIPTION
**Pull request by mantariksh**
## Problem

Upstream applications rely on this package to interact with the MyInfo API. However, if those applications are implemented in TypeScript, they have to write their own types for MyInfo, since this package does not provide any types.

## Solution

Implement this package in TypeScript.

### Build changes
A `build` directory was added to contain the compiled files and type declarations. The `main` field in `package.json` was changed to point to the compiled `build/index.js`, and a `types` field was added to point to `build/index.d.ts`.

A `.npmignore` file was added so files in `src` do not get published.

Linting was added for TypeScript.

The `test` script now includes a build step before the tests are run.

### Code changes
Most of the changes were purely to add types, but there were a few code changes as well:
- According to the docs, `mode` is optional in the constructor and `'prod'` would be used if it was not provided. However, the code did not contain any default for `mode`, so this default was added.
- The `jweResponse` is now passed straight into the `decrypt` function from `node-jose`. There is no need for us to do the parsing since [the library does it for us](https://github.com/cisco/node-jose/blob/1f86e23fc69063aea56974aa7247c2400b7dd50c/lib/jwe/decrypt.js#L71).

### Backwards compatibility
No changes were made to the API, so these changes are backwards-compatible.

## Other improvements
Github Actions were added to automate testing.

### CODE REVIEW

File: src/MyInfoGovClient.class.ts [hunk 1]
```diff
@@ -1,14 +1,42 @@
...
+import axios from 'axios'
+import { IPersonBasic } from './myinfo-types'
 
-const BASE_URL = {
-  dev: 'https://myinfosgstg.api.gov.sg/gov/dev/v1/',
-  stg: 'https://myinfosgstg.api.gov.sg/gov/test/v2/',
-  prod: 'https://myinfosg.api.gov.sg/gov/v2/',
+enum Mode {
+  Dev = 'dev',
+  Staging = 'sg',
```
Comment (liangyuanruo) [inline src/MyInfoGovClient.class.ts:11]:
```suggestion
  Staging = 'stg',
```

File: src/MyInfoGovClient.class.ts [hunk 2]
```diff
@@ -94,33 +129,33 @@ class MyInfoGovClient {
...
-   *    @param  {string} personRequest.uinFin - NRIC number
-   *    @param  {Array<string>} [personRequest.requestedAttributes] - Array of
+   *    @param personRequest.uinFin - NRIC number
+   *    @param [personRequest.requestedAttributes] - Array of
    *    attributes of person to request from MyInfo. Will query all fields if
    *    not provided, or if it is an empty list.
-   *    @param  {int} [personRequest.txnNo] - Optional transaction number
-   *    @param  {string} [singpassEserviceId] - Optional Singpass eService ID.
+   *    @param [personRequest.txnNo] - Optional transaction number
+   *    @param [singpassEserviceId] - Optional Singpass eService ID.
```
Comment (liangyuanruo) [inline src/MyInfoGovClient.class.ts:155]:
```suggestion
   *    @param [singpassEserviceId] - Optional Singpass eService ID that if provided, overrides the default eService ID.
```

File: src/MyInfoGovClient.class.ts [hunk 3]
```diff
@@ -387,8 +422,13 @@ class MyInfoGovClient {
    *          }
    *        }
    */
-  getPersonBasic ({ uinFin, requestedAttributes, txnNo, singpassEserviceId: seId }) {
-    if (!requestedAttributes || !requestedAttributes.length > 0) {
+  getPersonBasic ({ uinFin, requestedAttributes, txnNo, singpassEserviceId: seId }: {
```
Comment (liangyuanruo) [inline src/MyInfoGovClient.class.ts:425]: probably a good idea to create an interface for the argument here?

File: src/myinfo-types.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+interface IMyInfoMetadata {
+    lastUpdated: string
+    source: MyInfoSource
+    classification: 'C'
+}
+
+interface IMyInfoBasicField extends IMyInfoMetadata {
+    value: string
```
Comment (liangyuanruo) [inline src/myinfo-types.ts:21]: one of the MyInfo fields has a nasty bug where the value key is missing. it was reported 1-2 years back but we've not gotten a confirmation that it's been fixed. let me check which field it is and get back to you tomorrow.

```typescript
    value?: string
```

File: src/MyInfoGovClient.class.ts [hunk 5]
```diff
@@ -387,8 +422,13 @@ class MyInfoGovClient {
...
+    uinFin: string,
+    requestedAttributes?: string[],
+    txnNo?: string,
+    singpassEserviceId?: string
+  }): Promise<IPersonBasic> {
+    if (!requestedAttributes || requestedAttributes.length === 0) {
       requestedAttributes = ALL_ATTRIBUTES.personBasic
```
Comment (liangyuanruo) [inline src/MyInfoGovClient.class.ts:432]: convert ALL_ATTRIBUTES into an enum? too

File: src/myinfo-types.ts [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+    uinFin: string
+}
+
+// Check that IPersonBasic includes all keys from MyInfoAttribute
+type IPersonBasicCheck = Exclude<MyInfoAttribute, keyof IPersonBasic>
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+type VerifyIPersonBasicCheck<
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  Missing extends never = IPersonBasicCheck,
+> = never
```
Comment (liangyuanruo) [inline src/myinfo-types.ts:159]: clever

**Review by liangyuanruo — APPROVED** [review]
can we test this on staging before publishing? you may need to temporarily commit the build folder so that formsg staging can install from the github branch directly.

File: src/myinfo-types.ts (hunk 6 above)
Comment (mantariksh) [reply src/myinfo-types.ts:159]: https://stackoverflow.com/questions/51829842/how-to-force-interface-to-implement-keys-of-enum-in-typescript-3-0

**Comment by mantariksh** [comment]
hey @liangyuanruo, requesting another review due to the following changes:
- After much wailing and gnashing of teeth, I decided to abandon `axios` and go back to the `request` library. I tried console-logging the HTTP request in both cases, and the `Authorization` header was exactly the same. The MyInfo docs were also not very helpful because they didn't seem to line up with the existing code, probably because the examples are for V3 and not V2.
- I added `homeno` to the list of default attributes to request when no attribute list is specified. Not sure why it was left out in the first place.

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+          node-version: "12.x"
+      - name: Cache Node.js modules
+        uses: actions/cache@v2
+        with:
+          # npm cache files are stored in `~/.npm` on Linux/macOS
+          path: ~/.npm
+          key: ${{ runner.OS }}-node-${{ hashFiles('**/package-lock.json') }}
+          restore-keys: |
+            ${{ runner.OS }}-node-
+            ${{ runner.OS }}-
```
Comment (liangyuanruo) [inline .github/workflows/ci.yml:23]: Why does runner.OS appear twice?

**Review by liangyuanruo — APPROVED** [review]
lgtm! did you test against Mockpass too?

File: .github/workflows/ci.yml (hunk 7 above)
Comment (mantariksh) [reply .github/workflows/ci.yml:23]: when there's a cache miss for the exact key in `key`, the cache action searches for caches which match the list of `restore-keys`, hence the names in `restore-keys` are listed in decreasing order of specificity ([docs](https://docs.github.com/en/free-pro-team@latest/actions/guides/caching-dependencies-to-speed-up-workflows#example-using-the-cache-action))

**Pull request merged**